File: frontend/src/types/watering.ts
export type DueStatus = 'unrecorded' | 'due_today' | 'overdue'

export interface WateringPlantSummary {
  id: number
  name: string
  acquiredDate: string | null
  imageUrl: string | null
  wateringCycleDays: number
}

export interface PlantWateringState {
  plantId: number
  lastWateredAt: string | null
  nextWateringDate: string | null
  isDueToday: boolean
  hasWateredToday: boolean
  dueStatus: DueStatus | null
}

export interface WateringRecord {
  id: number
  plantId: number
  wateredAt: string
  createdAt: string
}

export interface UpcomingCareItem extends PlantWateringState {
  plant: WateringPlantSummary
}

export type UpcomingCareSectionKind = 'today' | 'tomorrow' | 'day_after_tomorrow' | 'future'

export interface UpcomingCareSection {
  date: string
  kind: UpcomingCareSectionKind
  items: UpcomingCareItem[]
}

export interface UpcomingCare {
  startDate: string
  days: number
  sections: UpcomingCareSection[]
}

export interface PlantWateringDetail extends PlantWateringState {
  history: WateringRecord[]
}

export interface WateringRecordCreateResult {
  record: WateringRecord
  state: PlantWateringDetail
}

export type WateringHeatmapLevel = 0 | 1 | 2 | 3 | 4

export interface WateringHeatmapRange {
  from: string
  to: string
}

export interface WateringHeatmapPlant {
  plantId: number
  name: string
}

export interface WateringHeatmapDay {
  date: string
  plantCount: number
  level: WateringHeatmapLevel
  plants: WateringHeatmapPlant[]
}

export interface WateringHeatmap {
  startDate: string
  endDate: string
  days: WateringHeatmapDay[]
}

File: backend/app/schemas/watering.py
from datetime import date, datetime, timezone
from typing import Literal

from pydantic import field_serializer
from sqlmodel import SQLModel

from app.schemas.plant import alias_config


DueStatus = Literal["unrecorded", "due_today", "overdue"]
UpcomingCareSectionKind = Literal["today", "tomorrow", "day_after_tomorrow", "future"]
WateringHeatmapLevel = Literal[0, 1, 2, 3, 4]


def serialize_utc_datetime(value: datetime) -> str:
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


class WateringSchema(SQLModel):
    model_config = alias_config


class WateringPlantSummaryRead(WateringSchema):
    id: int
    name: str
    image_url: str | None = None
    watering_cycle_days: int


class PlantWateringStateRead(WateringSchema):
    plant_id: int
    last_watered_at: datetime | None = None
    next_watering_date: date | None = None
    is_due_today: bool
    due_status: DueStatus | None = None

    @field_serializer("last_watered_at")
    def serialize_last_watered_at(self, value: datetime | None) -> str | None:
        if value is None:
            return None
        return serialize_utc_datetime(value)


class WateringRecordRead(WateringSchema):
    id: int
    plant_id: int
    watered_at: datetime
    created_at: datetime

    @field_serializer("watered_at", "created_at")
    def serialize_datetime(self, value: datetime) -> str:
        return serialize_utc_datetime(value)


class UpcomingCareItemRead(PlantWateringStateRead):
    plant: WateringPlantSummaryRead


class UpcomingCareSectionRead(WateringSchema):
    date: date
    kind: UpcomingCareSectionKind
    items: list[UpcomingCareItemRead]


class UpcomingCareRead(WateringSchema):
    start_date: date
    days: int
    sections: list[UpcomingCareSectionRead]


class PlantWateringDetailRead(PlantWateringStateRead):
    history: list[WateringRecordRead]


class WateringRecordCreateResult(WateringSchema):
    record: WateringRecordRead
    state: PlantWateringDetailRead


class WateringHeatmapPlantRead(WateringSchema):
    plant_id: int
    name: str


class WateringHeatmapDayRead(WateringSchema):
    date: date
    plant_count: int
    level: WateringHeatmapLevel
    plants: list[WateringHeatmapPlantRead]


class WateringHeatmapRead(WateringSchema):
    start_date: date
    end_date: date
    days: list[WateringHeatmapDayRead]
